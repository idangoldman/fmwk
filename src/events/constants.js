export const DOM_EVENTS_LIST = [
  ['clicks', ['click', 'dblclick']],
  ['keys', ['keyup', 'keydown', 'keypress']],
  ['mouse', ['mousedown', 'mouseenter', 'mouseleave', 'mousemove', 'mouseout', 'mouseover', 'mouseup']],
  ['window', ['contextmenu', 'resize', ['scroll', ['up|top', 'down|bottom', 'left', 'right']], 'select']],
  ['form', ['change', 'submit', 'reset', 'valid', 'invalid']],
  ['element', ['blur', 'focus', 'focusin', 'focusout']]
]
