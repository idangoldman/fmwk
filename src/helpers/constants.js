export const DOM_EVENTS_LIST = [
  ['clicks', ['click', 'dblclick']],
  ['keys', ['keyup', 'keydown', 'keypress']],
  ['mouse', ['mousedown', 'mouseenter', 'mouseleave', 'mousemove', 'mouseout', 'mouseover', 'mouseup']],
  ['window', ['contextmenu', 'resize', 'scroll', 'select']],
  ['form', ['change', 'submit', 'reset']],
  ['element', ['blur', 'focus', 'focusin', 'focusout']]
];

export const STORAGE_EVENTS_LIST = [
  ['change', ['remove', 'set', 'clear', 'empty']],
  'get'
];
