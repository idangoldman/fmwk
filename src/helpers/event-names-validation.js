/**
 * Match event names against given event names list.
 * @param  {String}      eventNames     String of event names separated by space.
 * @param  {EventsList}  eventNamesList Nested array of event names.
 * @return {Array}       Array of valid event names.
 * @throws When event name is not part of event names list.
 *
 * @example
 * const eventNames = 'click mouseover';
 *
 * eventNamesValidation(eventNames, DOM_EVENTS_LIST);
 * // ['click', 'mouseover']
 */
const eventNamesValidation = (eventNames = '', eventNamesList = []) => {
  const names = eventNames.split(' ');
  const list = eventNamesList.flat(3);

  for (const name of names) {
    if (!list.includes(name)) {
      throw new Error(`- No event was found with '${name}' name.`);
    }
  }

  return names;
};

export default eventNamesValidation;
