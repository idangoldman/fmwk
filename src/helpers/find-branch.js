const findBranch = (eventName = '', eventsList = []) => {
  return [].concat(eventsList).reduce((accumulator, currentValue, currentIndex, array) => {
    if (Array.isArray(currentValue)) {
      if (currentValue.includes(eventName)) {
        accumulator = accumulator.concat(eventName);
      } else {
        accumulator = accumulator.concat(
          findBranch(currentValue, eventName)
        );
      }
      array.splice(1);
    } else if (array.includes(eventName)) {
      accumulator = accumulator.concat(eventName);
      array.splice(1);
    } else {
      accumulator = accumulator.concat(currentValue);
    }

    return accumulator;
  }, []).reverse();
}

export default findBranch;
