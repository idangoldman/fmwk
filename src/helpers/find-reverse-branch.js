/**
 * @description
 * Searches an {@link EventsList} array for a specific item and returns the trail from it.
 *
 * @param  {String}     leaf  String value to be found in the array
 * @param  {EventsList} tree  Nested array of string values.
 * @return {Array}            Flat array that indicates the path to the leaf in reverse.
 *
 * @example
 *
 * const tree = [['branch_1', ['leaf_1_1', 'leaf_1_2']], ['branch_2', ['leaf_2_1']], 'leaf_0'];
 * const leaf = 'leaf_1_2';
 *
 * console.log(
 *   'The branch is:', findReverseBranch(leaf, tree)
 * );
 *
 * // ['leaf_1_2', 'branch_1']
 */

const findReverseBranch = (leaf = '', tree = []) => {
  const newBranch = [];

  if (tree.includes(leaf)) {
    newBranch.push(leaf);
  } else {
    for (const branch of tree) {
      if (!Array.isArray(branch)) continue;

      const [currentBranch, leafs] = branch;

      if (currentBranch === leaf) {
        newBranch.push(leaf);
      } else if (leafs.includes(leaf)) {
        newBranch.push(currentBranch, leaf);
      }
    }
  }

  return newBranch.reverse();
};

export default findReverseBranch;
