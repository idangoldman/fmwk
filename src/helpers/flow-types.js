// @flow

export type QuerySelectorType = string;

export type CallbackType = (?mixed) => void;

export type EventListType = Array<string | Array<string | Array<string>>>;
export type EventListenerArgsType = [string, CallbackType];
export type EventsStoreEventType = Map<string, Set<CallbackType>>;
export type EventsStoreType = Map<string, Map<string, Set<CallbackType>>>;
