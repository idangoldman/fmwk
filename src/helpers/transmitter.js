export default class Transmitter {
  constructor(eventNames = []) {
    this.STORE = new Map();
    this.EVENT_LIST = [].concat(eventNames);
  }

  on() {
  }

  off() {
  }

  once() {
  }

  emit() {
  }

  insert(key = '', eventNames = [], callback = undefined) {
    let events = this.STORE.get(key) || new Map();
    let names = this.validateEventNames(eventNames);
    let output = [];

    names.forEach((name) => {
      if (events.size && events.has(name)) {
        let event = events.get(name);

        if (!event.has(callback)) {
          event.add(callback);
          events.set(name, event);
        }
      } else {
        events.set(name, new Set([callback]));
      }

      output.push([name, callback]);
    });

    if (events.size) {
      this.STORE.set(key, events);
    }

    return output;
  }

  drop(key = '', eventNames = [], callback = undefined) {
    const events = this.STORE.get(key);
    const output = [];

    if (events) {
      if (eventNames.length) {
        const names = this.validateEventNames(eventNames);

        names.forEach((name) => {
          if (events.has(name)) {
            const event = events.get(name);

            if (!callback) {
              event.forEach(value => output.push([name, value]));
              event.clear();
            } else if (event.has(callback)) {
              output.push([name, callback]);
              event.delete(callback);
            }

            if (event.size) {
              events.set(name, event);
            } else {
              events.delete(name);
            }
          }
        });
      } else {
        events.forEach((callbacks, event) => {
          callbacks.forEach((_callback) => {
            output.push([event, _callback]);
          });
        });

        events.clear();
      }

      if (events.size) {
        this.STORE.set(key, events);
      } else {
        this.STORE.delete(key);
      }
    }

    return output;
  }

  validateEventNames(eventName = '') {
    let names = eventName.split(' ');

    names.every(name => {
      if (!this.findEventPath(this.EVENT_LIST, name).includes(name)) {
        throw new Error(`- No event was found with '${name}' name.`);
      }
    });

    return names;
  }

  findEventPath(eventsList, eventName) {
    return [].concat(eventsList).reduce((accumulator, currentValue, currentIndex, array) => {

      if (Array.isArray(currentValue)) {
        if (currentValue.includes(eventName)) {
          accumulator = accumulator.concat(eventName);
        } else {
          accumulator = accumulator.concat(
            this.findEventPath(currentValue, eventName)
          );
        }
        array.splice(1);
      } else if (array.includes(eventName)) {
        accumulator = accumulator.concat(eventName);
        array.splice(1);
      } else {
        accumulator = accumulator.concat(currentValue);
      }

      return accumulator;
    }, []).reverse();
  }
}
