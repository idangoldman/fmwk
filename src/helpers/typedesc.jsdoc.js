/**
* DOM instance of an element.
* @typedef {Object} HTMLElement
*/

/**
* A CSS selector string
* @typedef {String} QuerySelector
*/

/**
 *
 * A nested list of 3 levels deep, with a unique structure.
 * @typedef {Array} EventsList
 * @example
 * // [[key, [...values]], ...values]
 * const eventsList = [
 *  ['keys', ['keypress', 'keyup', 'keydown']],
 *  ['clicks', ['click', 'dblclick']],
 *  'resize',
 *  'reset'
 * ];
 */
