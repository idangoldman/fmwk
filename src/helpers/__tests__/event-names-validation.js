import eventNamesValidation from '/helpers/event-names-validation';
import { DOM_EVENTS_LIST } from '/helpers/constants';

describe('eventNamesValidation helper function tested', () => {
  test('Should be able validate event name', () => {
    const eventName = 'click';
    const match = ['click'];

    expect(eventNamesValidation(eventName, DOM_EVENTS_LIST)).toEqual(match);
  });

  test('Should be able validate multiple event names', () => {
    const eventName = 'click mouseover';
    const match = ['click', 'mouseover'];

    expect(eventNamesValidation(eventName, DOM_EVENTS_LIST)).toEqual(match);
  });

  test('Should not validate event names', () => {
    const eventName = 'clicked';
    const callback = () => eventNamesValidation(eventName, DOM_EVENTS_LIST);

    expect(callback).toThrowErrorMatchingSnapshot();
  });
});
