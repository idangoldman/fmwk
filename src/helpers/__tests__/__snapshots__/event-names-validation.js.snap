// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventNamesValidation helper function tested Should not be able validate event names 1`] = `"- No event was found with 'clicked' name."`;
